src/main/java/uk/gov/hmcts/ccd/domain/service/callbacks/CallbackService.java

```diff
@@ -185,13 +152,46 @@ private void sleep(final Integer timeout) {
         }
         final HttpEntity requestEntity = new HttpEntity(callbackRequest, httpHeaders);
 
-        StopWatch sw = new StopWatch();
-        sw.start();
-        ResponseEntity<T> exchange = restTemplate.exchange(url, HttpMethod.POST, requestEntity, clazz);
-        sw.stop();
-        LOG.info("CallbackExecutionTime={} caseType={} event={} url={}", sw.getTotalTimeMillis(), caseTypeId, eventId, url);
-        return ofNullable(exchange);
+        Future<ResponseEntity> future = executorService.submit(() -> restTemplate.exchange(url, HttpMethod.POST, requestEntity, clazz));
+
+        return handleResponse(url, timeout, future, caseTypeId, eventId);
+    }
+
+    @SuppressWarnings("javasecurity:S5145")
+    private <T> Optional<ResponseEntity<T>> handleResponse(final String url, final Integer timeout, final Future<ResponseEntity> future,
+                                                           final String caseTypeId, final String eventId) {
+        try {
+            StopWatch sw = new StopWatch();
+            sw.start();
+            Optional optionalResponseEntity = ofNullable(future.get(timeout, TimeUnit.SECONDS));
+            sw.stop();
+            LOG.info("CallbackExecutionTime={}ms caseType={} event={} url={}", sw.getTotalTimeMillis(), caseTypeId, eventId, url);
+            return optionalResponseEntity;
+        } catch (InterruptedException e) {
+            handleException("Task interrupted. ", url, e);
+            Thread.currentThread().interrupt();
+            throw new CallbackException(e.getMessage());
+        } catch (ExecutionException e) {
+            handleException("Execution exception. ", url, e);
+            if (e.getCause() instanceof RuntimeException) {
+                throw (RuntimeException)e.getCause();
+            } else {
+                return Optional.empty();
+            }
+        } catch (TimeoutException e) {
+            future.cancel(true);
+            handleException("Future timed out. ", url, e);
```

Comment:
is it possible to add an indication to the message that the callback will be retried?

Reply:
It will be logged anyway with the logging of `Trying url={} with timeout={}s` in the following retry attempt.